framework/include/utils/TransitionLayerTools.h

```diff
@@ -0,0 +1,210 @@
+ * @param linear_vec_y linear interpolation of y coordinates based on weighted index
+ * @param spline_vec_l spline interpolation of inter-node length based on weighted index
+ */
+void weightedInterpolator(const unsigned int vec_node_num,
+                          const std::vector<Point> boundary_points_vec,
+                          std::vector<Real> & vec_index,
+                          std::vector<Real> & wt,
+                          std::vector<Real> & index,
+                          const Real sigma,
+                          std::unique_ptr<LinearInterpolation> & linear_vec_x,
+                          std::unique_ptr<LinearInterpolation> & linear_vec_y,
+                          std::unique_ptr<SplineInterpolation> & spline_vec_l);
+
+/**
+ * Generates weighted surrogate index vectors on one side for points of a sublayer curve
+ * @param weighted_surrogate_index weighted surrogate index vector
+ * @param unweighted_surrogate_index unweighted surrogate index vector
+ * @param node_number_vec number of points/nodes on the sublayer curve
+ * @param index weighted index vector of the points on the side
+ * @param wt weight vector of the points on the side
+ * @param boundary_node_num number of points/nodes on the side
+ * @param i index of the sublayer
+ */
+void surrogateGenerator(std::vector<Real> & weighted_surrogate_index,
+                        std::vector<Real> & unweighted_surrogate_index,
+                        const std::vector<unsigned int> node_number_vec,
+                        const std::vector<Real> index,
+                        const std::vector<Real> wt,
+                        const unsigned int boundary_node_num,
+                        const unsigned int i);
+
+/**
+ * Decides whether all the Points of a vector of Points are on the XY plane
+ * @param vec_pts vector of points to be examined
+ * @return whether all the Points are on the XY plane
+ */
+bool isXYPlane(const std::vector<Point> vec_pts);
+
+/**
+ * Decide whether one of the input vector of Points needs to be flip to ensure correct transition
+ * layer shape
+ * @param vec_pts_1 first vector of points to be examined
+ * @param vec_pts_2 second vector of points to be examined
+ * @return whether one of the vectors needs to be flipped to ensure correct transition layer shape
+ */
+bool needFlip(const std::vector<Point> vec_pts_1, const std::vector<Point> vec_pts_2);
+
+/**
+ * Decides whether a boundary of a given mesh works with the algorithm used in this class.
+ * @param mesh input mesh that contains the boundary to be examined
+ * @param max_node_radius the maximum radius of the nodes on the
+ * boundary
+ * @param invalid_type help distinguish different types of invalid boundaries
+ * @param origin_pt origin position of the given mesh (used for azimuthal angle calculation)
+ * @param bid ID of the boundary to be examined
+ * @return whether the boundary works with the algorithm
+ */
+bool isBoundaryValid(ReplicatedMesh & mesh,
```

Comment:
These are similar to the methods in `PeripheralRingMeshGenerator`. Will consolidate them in either this PR or #20638

Reply:
Done